sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/model/PendingTransaction.kt

```diff
@@ -0,0 +1,143 @@
+    val rawTransactionId: FirstClassByteArray?
+) : SignedTransaction, Transaction
+
+// Note there are some commented out methods which aren't being removed yet, as they might be needed before the
+// Roomoval draft PR is completed
+
+// fun PendingTransaction.isSameTxId(other: MinedTransaction) =
+//     rawTransactionId == other.rawTransactionId
+//
+// fun PendingTransaction.isSameTxId(other: PendingTransaction) =
+//     rawTransactionId == other.rawTransactionId
+
+fun PendingTransaction.hasRawTransactionId() = rawTransactionId?.byteArray?.isEmpty() == false
+
+fun PendingTransaction.isCreating() =
+    raw.byteArray.isNotEmpty() && submitAttempts <= 0 && !isFailedSubmit() && !isFailedEncoding()
+
+fun PendingTransaction.isCreated() =
+    raw.byteArray.isNotEmpty() && submitAttempts <= 0 && !isFailedSubmit() && !isFailedEncoding()
+
+fun PendingTransaction.isFailedEncoding() = raw.byteArray.isNotEmpty() && encodeAttempts > 0
+
+fun PendingTransaction.isFailedSubmit(): Boolean {
+    return errorMessage != null || (errorCode != null && errorCode < 0)
+}
+
+fun PendingTransaction.isFailure(): Boolean {
+    return isFailedEncoding() || isFailedSubmit()
+}
+
+fun PendingTransaction.isCancelled(): Boolean {
+    return cancelled > 0
+}
+
+fun PendingTransaction.isMined(): Boolean {
+    return minedHeight != null
+}
+
+fun PendingTransaction.isSubmitted(): Boolean {
+    return submitAttempts > 0
+}
+
+@Suppress("ReturnCount")
+fun PendingTransaction.isExpired(
+    latestHeight: BlockHeight?,
+    saplingActivationHeight: BlockHeight
+): Boolean {
+    val expiryHeightLocal = expiryHeight
+
+    if (latestHeight == null || expiryHeightLocal == null) {
+        return false
+    }
+    // TODO: test for off-by-one error here. Should we use <= or <
+    if (latestHeight.value < saplingActivationHeight.value || expiryHeightLocal < saplingActivationHeight) {
+        return false
+    }
+
+    return expiryHeightLocal < latestHeight
+}
+
+private const val EXPIRY_BLOCK_COUNT = 100
+
+// if we don't have info on a pendingtx after 100 blocks then it's probably safe to stop polling!
+@Suppress("ReturnCount")
+fun PendingTransaction.isLongExpired(
+    latestHeight: BlockHeight?,
+    saplingActivationHeight: BlockHeight
+): Boolean {
+    val expiryHeightLocal = expiryHeight
+
+    if (latestHeight == null || expiryHeightLocal == null) {
+        return false
+    }
+
+    if (latestHeight.value < saplingActivationHeight.value || expiryHeightLocal < saplingActivationHeight) {
+        return false
+    }
+    return (latestHeight.value - expiryHeightLocal.value) > EXPIRY_BLOCK_COUNT
+}
+
+private const val ERROR_CODE_MARKED_FOR_DELETION = -9090
+
+fun PendingTransaction.isMarkedForDeletion(): Boolean {
+    return rawTransactionId == null && (errorCode ?: 0) == ERROR_CODE_MARKED_FOR_DELETION
+}
+
+private val smallThreshold = 30.minutes
+private val hugeThreshold = 30.days
+
+fun PendingTransaction.isSafeToDiscard(): Boolean {
```

Comment:
I know this maintains current functionality. But it should be important to revisit deletion of submission failures without consent of the user.